src/Compilers/CSharp/Test/Syntax/Parsing/RoleParsingTests.cs

```diff
@@ -0,0 +1,1206 @@
+                    {
+                        N(SyntaxKind.ColonToken);
+                        N(SyntaxKind.SimpleBaseType);
+                        {
+                            N(SyntaxKind.IdentifierName);
+                            {
+                                N(SyntaxKind.IdentifierToken, "U");
+                            }
+                        }
+                    }
+                    N(SyntaxKind.OpenBraceToken);
+                    N(SyntaxKind.CloseBraceToken);
+                }
+                M(SyntaxKind.CloseBraceToken);
+            }
+            N(SyntaxKind.EndOfFileToken);
+        }
+        EOF();
+    }
+
+    [Theory, CombinatorialData]
+    public void RoleParsing_BeforeTopLevelStatement(bool isExtension)
+    {
+        var keyword = isExtension ? "extension" : "role";
+        var text = $$"""
+{{keyword}} R : U { }
+Write();
+""";
+
+        UsingTree(text,
+            // (2,1): error CS8803: Top-level statements must precede namespace and type declarations.
+            // Write();
+            Diagnostic(ErrorCode.ERR_TopLevelStatementAfterNamespaceOrType, "Write();").WithLocation(2, 1)
+            );
+
+        N(SyntaxKind.CompilationUnit);
+        {
+            N(isExtension ? SyntaxKind.ExtensionDeclaration : SyntaxKind.RoleDeclaration);
+            {
+                N(isExtension ? SyntaxKind.ExtensionKeyword : SyntaxKind.RoleKeyword);
+                N(SyntaxKind.IdentifierToken, "R");
+                N(SyntaxKind.BaseList);
+                {
+                    N(SyntaxKind.ColonToken);
+                    N(SyntaxKind.SimpleBaseType);
+                    {
+                        N(SyntaxKind.IdentifierName);
+                        {
+                            N(SyntaxKind.IdentifierToken, "U");
+                        }
+                    }
+                }
+                N(SyntaxKind.OpenBraceToken);
+                N(SyntaxKind.CloseBraceToken);
+            }
+            N(SyntaxKind.GlobalStatement);
+            {
+                N(SyntaxKind.ExpressionStatement);
+                {
+                    N(SyntaxKind.InvocationExpression);
+                    {
+                        N(SyntaxKind.IdentifierName);
+                        {
+                            N(SyntaxKind.IdentifierToken, "Write");
+                        }
+                        N(SyntaxKind.ArgumentList);
+                        {
+                            N(SyntaxKind.OpenParenToken);
+                            N(SyntaxKind.CloseParenToken);
+                        }
+                    }
+                    N(SyntaxKind.SemicolonToken);
+                }
+            }
+            N(SyntaxKind.EndOfFileToken);
+        }
+        EOF();
+    }
+
+    [Fact]
+    public void SyntaxFactory_TypeDeclaration()
+    {
+        Assert.Throws<InvalidOperationException>(() => SyntaxFactory.TypeDeclaration(SyntaxKind.RoleDeclaration, "R"));
```

Comment:
>InvalidOperationException [](http://example.com/codeflow?start=22&length=25)

Why do these calls to `SyntaxFactory.TypeDeclaration()` throw an exception but the corresponding call with `SyntaxKind.RecordDeclaration` does not?

Reply:
Unlike other type declarations, there is no way to produce a working syntax when only given a name. Roles require an underlying type.
I'm open to suggestions.